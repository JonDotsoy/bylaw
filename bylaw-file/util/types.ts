import { PlistObject, PlistValue } from 'plist';
import { isRegExp } from 'util/types';

// help

export type Obj = { [key: string]: any };

export const isBuffer = (value: PlistValue): value is Buffer => value instanceof Buffer;
export const isDate = (value: PlistValue): value is Date => value instanceof Date
export const isString = (value: PlistValue): value is string => typeof value === 'string'
export const isNumber = (value: PlistValue): value is number => typeof value === 'number'
export const isBoolean = (value: PlistValue): value is boolean => typeof value === 'boolean'
export const isArray = (value: any): value is any[] => Array.isArray(value)

export const isObject = (value: any): value is Obj => value !== null &&
  typeof value === 'object' &&
  !Array.isArray(value) &&
  !isBuffer(value) &&
  !isDate(value)

export const isPlistArray = (value: PlistValue): value is PlistObject[] => Array.isArray(value)
export const isPlistObject = (value: PlistValue): value is PlistObject => typeof value === 'object'
  && value !== null
  && !isPlistArray(value)
  && !isBuffer(value)
  && !isDate(value)

export const isStringOrRegExp = (value: any) => isString(value) || isRegExp(value);
