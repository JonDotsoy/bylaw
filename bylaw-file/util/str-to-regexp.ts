export const strToExp = (value: string | RegExp) => {
  if (value instanceof RegExp) return value;

  const newLocal_1 = Array.from(Buffer.from(value, 'utf-8'))
    .map(e =>
      e == 0x08
        ? Array.from(Buffer.from('\\b', 'utf-8'))
        : e
    )
    .flat();

  const newLocal = Buffer.from(newLocal_1).toString('utf-8');

  return new RegExp(newLocal);
};
