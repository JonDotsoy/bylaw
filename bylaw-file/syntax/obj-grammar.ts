import { AsyncLocalStorage } from 'async_hooks'
import { inspect, InspectOptions } from 'util';
import { isObject, Obj } from '../util/types';

type MapIndex = Map<string, any>;
type ListPatterns = Set<any>;

export const grammarAsyncLocalStorage = new AsyncLocalStorage<{
  indexes: MapIndex,
  patterns: ListPatterns,
}>();

const eachIndex = <T>(v: T): T => {
  const indexes = grammarAsyncLocalStorage.getStore()!.indexes
  Object.entries(v).forEach(([key, value]) => {
    indexes.set(`#${key}`, value)
  })
  return v;
};

const toPattern = <T extends Obj>(v: T) => {
  const indexes = grammarAsyncLocalStorage.getStore()!.indexes;
  if (isObject(v) && 'include' in v) {
    return {
      ...v,
      toRef: () => indexes.get(v.include),
    }
  }
}

export const createGrammar = <T>(v: () => T): T => grammarAsyncLocalStorage.run({
  indexes: new Map<string, any>(),
  patterns: new Set<any>(),
}, () => v());


export class Repository {
  #entry: any;
  [k: string]: any;

  private constructor(entry: any) {
    this.#entry = entry;
    const { indexes } = grammarAsyncLocalStorage.getStore()!;
    for (const [key, value] of Object.entries(entry)) {
      if (typeof key === 'string') {
        this[key] = value;
        indexes.set(`#${key}`, value);
      }
    }
  }

  [inspect.custom](depth: number, options: InspectOptions) {
    const inspectConfig = inspectConfigurable.getStore()

    if (inspectConfig?.type === 'ts') {
      return `Repository.create(${inspect(this.#entry, options)})`
    }

    return `Repository ${inspect(this.#entry, options)}`
  }

  static create(entry: any) {
    return new Repository(entry);
  }
}

export enum EntryType {
  inlinePattern = 'inline-pattern',
  include = 'include',
  match = 'match',
  repository = 'repository',
}


type Entry<T extends EntryType> =
  T extends EntryType.inlinePattern ? {
    begin: RegExp
    beginCaptures: { [propName: `${number}`]: { name: string } }
    patterns: Pattern<any>[]
    end: RegExp
    endCaptures: { [propName: `${number}`]: { name: string } }
  }
  : T extends EntryType.include ? { include: string }
  : T extends EntryType.repository ? {
    patterns: Pattern<any>[]
  }
  : T extends EntryType.match ? {
    match: RegExp
    name: string
  }
  : null;

export class Pattern<T extends EntryType> {
  #type: T
  #entry: Entry<T>;
  #indexes: MapIndex
  #patterns: ListPatterns;

  private constructor(
    type: T,
    entry: any,
  ) {
    this.#entry = entry;
    this.#type = type;
    const { indexes, patterns } = grammarAsyncLocalStorage.getStore()!;
    this.#indexes = indexes;
    this.#patterns = patterns;
    patterns.add(this);
  }

  getIndexes() { return this.#indexes }
  getPatterns() { return this.#patterns }

  * deepPatterns(): Generator<Pattern<any>> {
    if ('patterns' in this.#entry) {
      for (const pattern of this.#entry.patterns) {
        if (Pattern.isPatternType(pattern, EntryType.include)) {
          for (const pattern2 of this.getIndexes().get(pattern.#entry.include).patterns) {
            yield pattern2;
          }
        }
      }
    }
  }

  match(value: string, indexOf = 0): any {
    if (Pattern.isPatternType(this, EntryType.repository)) {
      const patterns: Pattern<any>[] = this.#entry.patterns;

      for (const pattern of patterns) {
        const result = pattern.match(value);
        if (result) {
          return result;
        }
      }

      return null;
    }

    if (Pattern.isPatternType(this, EntryType.include)) {
      const patterns = this.getIndexes().get(this.#entry.include).patterns;

      for (const pattern of patterns) {
        const result = pattern.match(value, indexOf);
        if (result) {
          return result;
        }
      }

      return null;
    }

    if (Pattern.isPatternType(this, EntryType.inlinePattern)) {
      const { begin, beginCaptures } = this.#entry;
      const beginExp = new RegExp(`^(?<prefix>(?:\\n|\\s)*)${begin.source}`);

      const resultMatch = beginExp.exec(value);

      if (resultMatch) {
        const postEnd = indexOf + resultMatch[0].length;
        let postEndResult = postEnd;

        const { end, endCaptures, } = this.#entry;
        const childs: any[] = [];

        const patterns = Array.from(this.deepPatterns());

        for (let index = 0; index < patterns.length; index++) {

          const pattern = patterns[index];
          const result = pattern.match(value.slice(postEndResult), postEndResult);
          if (result) {
            childs.push(result);
            postEndResult = result.end;
            index = 0;
          }
        }

        return {
          tokens: Object.entries(beginCaptures).map(([key, value]) => {
            return {
              name: value.name,
              value: resultMatch[Number(key) + 1],
            }
          }),
          start: resultMatch.index + indexOf + resultMatch.groups!.prefix.length,
          end: postEnd,
          childs,
        }
      }

      return null;
    }

    if (Pattern.isPatternType(this, EntryType.match)) {
      const matchExp = new RegExp(`^(?<prefix>(?:\\n|\\s)*)${this.#entry.match.source}`);
      const matchResult = matchExp.exec(value);
      if (matchResult) {
        const postEnd = indexOf + matchResult[0].length;
        return {
          tokens: [{
            name: this.#entry.name,
            value: matchResult[2],
          }],
          // value,
          // matchResult,
          // source: matchExp.source,
          // indexOf,
          start: indexOf + matchResult.index + matchResult.groups!.prefix.length,
          end: postEnd,
        }
      }
      // console.log(this.#entry.name, matchExp.source, JSON.stringify(value), matchResult);
      return null;
      throw new Error('Not implemented');
    }

    throw new Error(`Unknown type ${this.#type}`);
  }

  [inspect.custom](depth: number, options: InspectOptions) {
    const inspectConfig = inspectConfigurable.getStore()

    if (inspectConfig?.type === 'ts') {
      return `Pattern.create(${inspect(this.#type)}, ${inspect(this.#entry, options)})`
    }

    return `Pattern <${this.#type}> ${inspect(Object.fromEntries(Object.entries(this)), options)}`
  }

  static isPatternType<T extends EntryType>(pattern: Pattern<any>, type: T): pattern is Pattern<T> {
    return pattern.#type === type;
  }

  static create(
    type: EntryType,
    entries: any,
  ) {
    return new Pattern(type, entries);
  }

  static isPattern(value: any): value is Pattern<any> {
    return value instanceof Pattern;
  }
}

type InspectConfig = {
  type: 'ts'
}

const inspectConfigurable = new AsyncLocalStorage<InspectConfig>();

export const inspectToTS = (value: Pattern<any>) => inspectConfigurable.run(
  {
    type: 'ts',
  },
  () => `createGrammar(() => ${inspect(value, {
    depth: null,
    maxArrayLength: null,
    maxStringLength: null,
    showHidden: false,
  })})`
)
