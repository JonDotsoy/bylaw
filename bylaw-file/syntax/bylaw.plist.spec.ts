import { parse } from './bylaw.plist'
import fs from 'fs'
import { inspectToTS, Pattern } from './obj-grammar'
import { inspect } from 'util'


it('parse', () => {
  const parsed = parse(fs.readFileSync(`${__dirname}/syntax.plist`, 'utf-8'))

  expect(parsed).toBeInstanceOf(Pattern)
})

it('stringify to ts', () => {
  const parsed = parse(fs.readFileSync(`${__dirname}/syntax.plist`, 'utf-8'))

  expect(inspectToTS(parsed)).toMatchSnapshot()
})


it('should match pattern', () => {
  const parsed = parse(fs.readFileSync(`${__dirname}/syntax.plist`, 'utf-8'));

  const source = `
    version = "1.0";
  `;

  const match = parsed.match(source);

  console.log(inspect(match, {
    depth: null,
    maxArrayLength: null,
  }));
})

