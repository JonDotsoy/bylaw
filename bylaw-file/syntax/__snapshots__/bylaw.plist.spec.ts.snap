// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringify to ts 1`] = `
"createGrammar(() => Pattern.create('repository', {
  fileTypes: [ 'bylaw' ],
  name: 'bylaw',
  patterns: [ Pattern.create('include', { include: '#main' }) ],
  scopeName: 'source.bylaw',
  uuid: '',
  repository: Repository.create({
    main: {
      patterns: [
        Pattern.create('include', { include: '#match' }),
        Pattern.create('include', { include: '#variable' })
      ]
    },
    allow_exception: {
      patterns: [
        Pattern.create('inline-pattern', {
          begin: /(\\\\ballow\\\\b)/,
          beginCaptures: { '1': { name: 'storage.type.bylaw' } },
          patterns: [ Pattern.create('include', { include: '#allow_exception__1' }) ],
          end: /(\\\\:)/,
          endCaptures: { '1': { name: 'storage.type.bylaw' } }
        }),
        Pattern.create('include', { include: '#inline_condition' }),
        Pattern.create('match', { match: /(.)/, name: 'invalid.illegal.bylaw' })
      ]
    },
    allow_exception__1: {
      patterns: [
        Pattern.create('match', { match: /(\\\\b[a-z_A-Z0-9]+)/, name: 'string.bylaw' }),
        Pattern.create('match', { match: /(,)/, name: 'storage.type.bylaw' }),
        Pattern.create('match', { match: /(.)/, name: 'invalid.illegal.bylaw' })
      ]
    },
    boolean: {
      patterns: [
        Pattern.create('inline-pattern', {
          begin: /(true|false)/,
          beginCaptures: { '1': { name: 'constant.other.bylaw' } },
          patterns: [ Pattern.create('include', { include: '#boolean__1' }) ],
          end: /(\\\\b)/,
          endCaptures: { '1': { name: 'storage.type.bylaw' } }
        })
      ]
    },
    boolean__1: {
      patterns: [
        Pattern.create('match', { match: /(.*)/, name: 'invalid.illegal.bylaw' })
      ]
    },
    inline_condition: {
      patterns: [
        Pattern.create('inline-pattern', {
          begin: /(\\\\bif)/,
          beginCaptures: { '1': { name: 'storage.type.bylaw' } },
          patterns: [ Pattern.create('include', { include: '#inline_condition__1' }) ],
          end: /(\\\\;)/,
          endCaptures: { '1': { name: 'storage.type.bylaw' } }
        })
      ]
    },
    inline_condition__1: {
      patterns: [
        Pattern.create('match', { match: /(\\\\b(or|and)\\\\b)/, name: 'storage.type.bylaw' }),
        Pattern.create('match', { match: /(==)/, name: 'storage.type.bylaw' }),
        Pattern.create('include', { include: '#string' }),
        Pattern.create('include', { include: '#boolean' }),
        Pattern.create('include', { include: '#number' }),
        Pattern.create('include', { include: '#selector' }),
        Pattern.create('match', { match: /(.)/, name: 'invalid.illegal.bylaw' })
      ]
    },
    match: {
      patterns: [
        Pattern.create('inline-pattern', {
          begin: /(\\\\s*\\\\bmatch)/,
          beginCaptures: { '1': { name: 'storage.type.bylaw' } },
          patterns: [ Pattern.create('include', { include: '#match__1' }) ],
          end: /(^(?=.{0,1})(?:|))/,
          endCaptures: { '1': { name: 'variable.bylaw' } }
        })
      ]
    },
    match__1: {
      patterns: [
        Pattern.create('include', { include: '#router_pattern' }),
        Pattern.create('inline-pattern', {
          begin: /(\\\\s*\\\\{)/,
          beginCaptures: { '1': { name: 'storage.type.bylaw' } },
          patterns: [ Pattern.create('include', { include: '#match__2' }) ],
          end: /(\\\\s*\\\\})/,
          endCaptures: { '1': { name: 'storage.type.bylaw' } }
        })
      ]
    },
    match__2: {
      patterns: [
        Pattern.create('include', { include: '#allow_exception' }),
        Pattern.create('match', { match: /([^\\\\s])/, name: 'invalid.illegal.bylaw' })
      ]
    },
    number: {
      patterns: [
        Pattern.create('match', {
          match: /(\\\\s*(\\\\b|-)\\\\d+(_\\\\d+)*(\\\\.\\\\d+(_\\\\d+)*)?)/,
          name: 'constant.numeric.bylaw'
        })
      ]
    },
    router_pattern: {
      patterns: [
        Pattern.create('inline-pattern', {
          begin: /([^\\\\/a-zA-Z0-9_-\\\\{\\\\}])/,
          beginCaptures: { '1': { name: 'string.bylaw' } },
          patterns: [ Pattern.create('include', { include: '#router_pattern__1' }) ],
          end: /([^\\\\/a-zA-Z0-9_-\\\\{\\\\}])/,
          endCaptures: { '1': { name: 'string.bylaw' } }
        })
      ]
    },
    router_pattern__1: {
      patterns: [
        Pattern.create('match', { match: /([\\\\/a-zA-Z0-9_-])/, name: 'string.bylaw' }),
        Pattern.create('inline-pattern', {
          begin: /(\\\\{)/,
          beginCaptures: { '1': { name: 'storage.type.bylaw' } },
          patterns: [ Pattern.create('include', { include: '#router_pattern__2' }) ],
          end: /(\\\\})/,
          endCaptures: { '1': { name: 'storage.type.bylaw' } }
        })
      ]
    },
    router_pattern__2: {
      patterns: [
        Pattern.create('match', { match: /(\\\\b[a-z_A-Z0-9]+)/, name: 'variable.bylaw' }),
        Pattern.create('match', { match: /([^\\\\s]*)/, name: 'invalid.illegal.bylaw' })
      ]
    },
    selector: {
      patterns: [
        Pattern.create('match', {
          match: /(\\\\b[\\\\w\\\\x{0024}_]+(\\\\s*(\\\\.\\\\s*[\\\\w\\\\x{0024}_]+|\\\\[\\\\s*(\\".*?\\"|'.*?'|-?\\\\d+(_\\\\d+)*)\\\\s*\\\\]))*)/,
          name: 'variable.bylaw'
        })
      ]
    },
    string: {
      patterns: [
        Pattern.create('include', { include: '#string_doble_quote' }),
        Pattern.create('include', { include: '#string_simple_quote' })
      ]
    },
    string_doble_quote: {
      patterns: [
        Pattern.create('inline-pattern', {
          begin: /(\\\\\\")/,
          beginCaptures: { '1': { name: 'string.bylaw' } },
          patterns: [ Pattern.create('include', { include: '#string_doble_quote__1' }) ],
          end: /([^\\\\x{005c}\\\\x{0022}]\\\\\\")/,
          endCaptures: { '1': { name: 'string.bylaw' } }
        })
      ]
    },
    string_doble_quote__1: {
      patterns: [
        Pattern.create('match', { match: /(\\\\\\\\\\\\\\"|[^\\\\\\"])/, name: 'string.bylaw' })
      ]
    },
    string_simple_quote: {
      patterns: [
        Pattern.create('inline-pattern', {
          begin: /(\\\\s*\\\\')/,
          beginCaptures: { '1': { name: 'string.bylaw' } },
          patterns: [ Pattern.create('include', { include: '#string_simple_quote__1' }) ],
          end: /([^\\\\x{005c}\\\\x{0027}]\\\\')/,
          endCaptures: { '1': { name: 'string.bylaw' } }
        })
      ]
    },
    string_simple_quote__1: {
      patterns: [
        Pattern.create('match', { match: /(\\\\\\\\\\\\'|[^\\\\'])/, name: 'string.bylaw' })
      ]
    },
    variable: {
      patterns: [
        Pattern.create('inline-pattern', {
          begin: /(\\\\s*\\\\b[a-z_A-Z0-9]+)/,
          beginCaptures: { '1': { name: 'variable.bylaw' } },
          patterns: [ Pattern.create('include', { include: '#variable__1' }) ],
          end: /(\\\\s*\\\\;)/,
          endCaptures: { '1': { name: 'storage.type.bylaw' } }
        })
      ]
    },
    variable__1: {
      patterns: [
        Pattern.create('match', { match: /(\\\\s*\\\\=)/, name: 'storage.type.bylaw' }),
        Pattern.create('include', { include: '#string' }),
        Pattern.create('include', { include: '#number' }),
        Pattern.create('include', { include: '#boolean' }),
        Pattern.create('match', { match: /([^\\\\s])/, name: 'invalid.illegal.bylaw' })
      ]
    }
  })
}))"
`;
